### ISSUE TITLE
Potential new feature - automatic proxy identification?

### ISSUE THREAD

**Issue by MaximeAgez** [LCA-ActivityBrowser/activity-browser#1001]
I've recently developed a Python module matching words to a list of words through a machine learning algorithm (https://github.com/CIRAIG/IE_ML_mapping).
The way I am primarily using this module is to match different classifications together, but for Activity-browser I think it could be used as a proxy identification tool.

**_What the module does_**:
You give it a word (or list of words), say "Chocolate" and a target classification (e.g., CPC) and the module will automatically return a number of "guesses" for what "Chocolate" ressembles the most in the CPC classification. As you can see in the screenshot below, it is word association, and not simply word ressemblance (as in similar letters together).
![image](https://github.com/LCA-ActivityBrowser/activity-browser/assets/24285636/f7143e10-1c44-4391-8f4b-57c76df45a2e)

**_How it could be used within AB_**:
Now if the target classification is ecoinvent3.8. If i were to look for "Chocolate" in the current filter of AB, I would not find anything. Say AB had an option where you could use the Python module, "Chocolate" would perhaps be associated to "cocoa bean" and "sugar". Granted these are not adequate proxies (on this particular example) and the practitioner should really look for an existing LCA of "Chocolate" in the literature, but that would be a start.

If the community finds it interesting I could help implement the feature, but I am not proficient enough in Qt to implement it myself (nor do I have time :x)

**Comment by marc-vdm** [comment]
Hey this sounds really cool! So we recently launched plugins with AB `2.8.0`, which allows people to write additional functionality for AB that they then use. This could be a really cool plugin to add to AB :).

If the UI side would not be very complicated (how many buttons, interaction etc), we could see if we can help from the AB side.

**Comment by MaximeAgez** [comment]
To calculate the similarity scores is not long. The long part is to treat the data of all ecoinvent activities/products. So this should be ran once and stored in a database somewhere.

**Comment by marc-vdm** [comment]
So how do the similarity scores work? I'd assume you need to calculate those for every word you search no? Anyway, as I said above, this could definitely be made into a third party plugin, if needed I'd be happy to think along with the UI and if I have time help write some Qt for it!

**Comment by MaximeAgez** [comment]
Yes, everytime a word/sentence is entered the code will generate a numpy array with multiple values representing characteristics of the sentence. Then it will compare the generated array to the arrays of all the ecoinvent processes and determine similarity scores based on that.

Which means that:
- Need to precalculate all the arrays for the 19000 processes of ecoinvent. Could include the description of the process which might give better matching (from what I understand the more words the better)
- Store these matrices, in a new proxy.db SQL database within the package maybe?
- Need to calculate the array each time a user would want to search a word/combination of words
- Need to match the generated array with the stored matrices

The last two steps are required for every use but should be fast. The two first steps are extremely long but need being ran only once.

I will try and look into the plugins when I have the time and come back if I encounter difficulties.

**Comment by marc-vdm** [comment]
Alright sounds great! If it's indeed just pre-calculating some values they can be stored in the package.

**Comment by bsteubing** [comment]
Hi Maxime,
as Marc has already said this is a cool idea and feature, but I am hesitant to include this into the AB for a number of reasons at the moment. One of the reasons is that in the mid-term we need to do some larger revisions of the code (e.g. migration to bw2.5) and that has priority of and should happen before adding new features and dependencies. However, if you happen to make a stand-alone tool from it where people can enter what they search for an get proxies, then we could point to that from our Wiki (or from within the AB even).
